src/bridge/cockpitdbusmachines.c

```diff
@@ -0,0 +1,451 @@
+    {
+      /* update settings for hostname */
+      g_assert (JSON_NODE_HOLDS_OBJECT(cur_props));
+      json_object_foreach_member (json_node_get_object (info), update_machine_property, json_node_get_object (cur_props));
+    }
+  else
+    /* create new entry for host name */
+    json_object_set_member (cur_config_obj, hostname, json_node_copy (info));
+
+  /* write back json file */
+  json_gen = json_generator_new ();
+  json_generator_set_root (json_gen, cur_config);
+  json_generator_set_pretty (json_gen, TRUE); /* bikeshed zone */
+  res = json_generator_to_file (json_gen, path, error);
+  g_free (path);
+  g_object_unref (json_gen);
+  json_node_free (cur_config);
+  return res;
+}
+
+static GVariant *
+machines_get_property (GDBusConnection *connection,
+                       const gchar *sender,
+                       const gchar *object_path,
+                       const gchar *interface_name,
+                       const gchar *property_name,
+                       GError **error,
+                       gpointer user_data)
+{
+  g_return_val_if_fail (property_name != NULL, NULL);
+
+  if (g_str_equal (property_name, "Machines"))
+    return get_machines ();
+  else
+    g_return_val_if_reached (NULL);
+}
+
+static void
+machines_method_call (GDBusConnection *connection,
+                      const gchar *sender,
+                      const gchar *object_path,
+                      const gchar *interface_name,
+                      const gchar *method_name,
+                      GVariant *parameters,
+                      GDBusMethodInvocation *invocation,
+                      gpointer user_data)
+{
+  if (g_str_equal (method_name, "Update"))
+    {
+      const gchar *filename, *hostname;
+      GVariant * info_v;
+      JsonNode *info_json = NULL;
+      GError *error = NULL;
+
+      g_variant_get (parameters, "(&s&s@a{sv})", &filename, &hostname, &info_v);
+      info_json = json_gvariant_serialize (info_v);
+      /* g_debug ("Updating %s for machine %s: %s", filename, hostname, g_variant_print (info_v, TRUE)); */
+      g_variant_unref (info_v);
+
+      if (update_machine (filename, hostname, info_json, &error))
+        g_dbus_method_invocation_return_value (invocation, NULL);
+      else
+        g_dbus_method_invocation_take_error (invocation, error);
+      json_node_free (info_json);
+    }
+  else
+    g_return_if_reached ();
+}
+
+/**
+ * update_properties:
+ * @user_data: GDBusConnection to which updates get sent
+ *
+ * Update the "Machines" property and send a PropertiesChanged signal with it.
+ * This gets called in reaction to changed *.json configuration files.
+ */
+static gboolean
+update_properties (gpointer user_data)
+{
+  GDBusConnection *connection = user_data;
+  GVariant *machines, *signal_value;
+  GVariantBuilder builder;
+  GError *error = NULL;
+
+  /* reset pending counter before we do any actual work, to avoid races */
+  pending_updates = 0;
+  machines = get_machines ();
```

Comment:
I'm not against using a property for Machines ... but we should be using the ```invalidated``` parameter of PropertiesChanged instead of automatically doing this expensive operation each time ... even if nobody's listening.